File: algorithms/code/leetcode/lc029_divide_two_integers/LC029DivideTwoIntegers.java
package leetcode.lc029_divide_two_integers;

/**
 * https://leetcode.com/problems/divide-two-integers/
 */
public final class LC029DivideTwoIntegers {
    public int divide(final int shortDividend, final int shortDivisor) {
        long dividend = shortDividend;
        long divisor = shortDivisor;
        if (divisor == 0) {
            return Integer.MAX_VALUE;
        }
        if (dividend == 0) {
            return 0;
        }
        long sign = (dividend / Math.abs(dividend)) * (divisor / Math.abs(divisor));
        dividend = Math.abs(dividend);
        divisor = Math.abs(divisor);
        long counter = 0;
        while (dividend >= divisor) {
            long subtractor = divisor;
            long incrementor = 1;
            while (dividend - subtractor - subtractor >= 0) {
                subtractor += subtractor;
                incrementor += incrementor;
            }
            dividend -= subtractor;
            counter += incrementor;
        }
        if (sign * counter > Integer.MAX_VALUE) {
            return Integer.MAX_VALUE;
        }
        return (int) (sign * counter);
    }
}

File: algorithms/code/leetcode/lc029_divide_two_integers/lc029_divide_two_integers.go
// Package lc029 implements https://leetcode.com/problems/divide-two-integers/
package lc029

import (
	"math"
)

func abs(a int64) int64 {
	if a < 0 {
		a = -a
	}
	return a
}

// Divide divides two integers without using multiplication, division, and mod operator
func Divide(dividend int, divisor int) int {
	if divisor == 0 {
		return math.MaxInt32
	}
	if dividend == 0 {
		return 0
	}
	longDividend := int64(dividend)
	longDivisor := int64(divisor)
	sign := (longDividend / abs(longDividend)) * (longDivisor / abs(longDivisor))
	longDividend = abs(longDividend)
	longDivisor = abs(longDivisor)
	var counter int64
	for longDividend >= longDivisor {
		subtractor := longDivisor
		incrementor := int64(1)
		for longDividend-subtractor-subtractor >= 0 {
			subtractor += subtractor
			incrementor += incrementor
		}
		longDividend -= subtractor
		counter += incrementor
	}
	result := sign * counter
	if result > math.MaxInt32 {
		return math.MaxInt32
	}
	return int(result)
}
